File: leetcode/py/p392-is-subsequence.py
# LeetCode 392. Is Subsequence.


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if len(s) == 0:
            return True
        i = 0
        for c in t:
            if c == s[i]:
                i += 1
                if i == len(s):
                    return True
        return False


test_cases = (
    ("abc", "ahbgdc", True),
    ("axc", "ahbgdc", False),
    ("", "", True),
    ("a", "a", True),
    ("a", "b", False),
)

solution = Solution()
for s, t, exp in test_cases:
    act = solution.isSubsequence(s, t)
    if exp != act:
        print(f"Fail. s: {s}, t: {t}, exp: {exp}, act: {act}")

File: leetcode/cpp/p392-is-subsequence.cpp
// LeetCode 392. Is Subsequence.

#include <iostream>

class Solution {
public:
    bool isSubsequence(std::string s, std::string t)
    {
        if (s.empty()) {
            return true;
        }
        size_t i = 0;
        for (const char c : t) {
            if (c == s[i]) {
                ++i;
                if (i == s.size()) {
                    return true;
                }
            }
        }
        return false;
    }
};

int main()
{
    struct TestCase {
        std::string s;
        std::string t;
        bool exp;
    };

    const TestCase test_cases[] = {
        { "abc", "ahbgdc", true },
        { "axc", "ahbgdc", false },
        { "", "ahbgdc", true },
        { "a", "a", true },
        { "a", "b", false },
    };

    Solution solution;
    for (const auto& tc : test_cases) {
        const auto act = solution.isSubsequence(tc.s, tc.t);
        if (tc.exp != act) {
            std::cerr << "fail, s: " << tc.s
                      << ", t: " << tc.t
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
    return 0;
}
